src/metabase/driver/sql_jdbc/execute.clj

```diff
@@ -259,10 +305,42 @@
           (.cancel stmt))))
     stmt))
 
-(defmethod execute-query! :sql-jdbc
+(defn- use-statement? [driver params]
+  (and (statement-supported? driver) (empty? params)))
+
+(defn- statement*
+  ^Statement [driver conn canceled-chan]
+  ;; if canceled-chan gets a message, cancel the Statement
+  (let [^Statement stmt (statement driver conn)]
+    (a/go
+      (when (a/<! canceled-chan)
+        (log/debug (trs "Query canceled, calling Statement.cancel()"))
+        (u/ignore-exceptions
+         (.cancel stmt))))
+    stmt))
+
+(defn- ^Statement statement-or-prepared-statement [driver conn sql params canceled-chan]
```

Comment:
Just noticed this. Tags on functions should go on the argslist, not on the function name. It should be 

```clj
(defn-  statement-or-prepared-statement ^Statement [driver conn sql params canceled-chan]
  ...)
```

Otherwise you could do something like
```clj
(with-open [stmt statement-or-prepared-statement]
  ...)
```

without the compiler catching it, because you've marked `statement-or-prepared-statement` itself as a `Statement`, rather than the result of a function call

Reply:
> Just noticed this. Tags on functions should go on the argslist, not on the function name. It should be
> 
> ```clojure
> (defn-  statement-or-prepared-statement ^Statement [driver conn sql params canceled-chan]
>   ...)
> ```
> 
> Otherwise you could do something like
> 
> ```clojure
> (with-open [stmt statement-or-prepared-statement]
>   ...)
> ```
> 
> without the compiler catching it, because you've marked `statement-or-prepared-statement` itself as a `Statement`, rather than the result of a function call

Opened #15102 to address these issues.